### PR TITLE
Improve handling of AndTypes on the LHS of subtype comparisons

### LINKED ISSUE #12077: Type inference failure for type variable in intersection type
## Compiler version

2.13.5 and  3.0.0-RC2

## Minimized code

This code compiles in 2.13.5  but not in 3.0.0-RC2.

```Scala
    trait Wrapper[K]
    trait Has0[T]
    
    def test[R](v: Wrapper[Has0[String] with R]):R = ???

    val zz:Wrapper[Has0[String] with Has0[Int]] = ???
    test(zz)
```

## Output

Error in 3.0.0-RC2

```scala
Found:    (zz : Wrapper[Has0[String] & Has0[Int]])
Required: Wrapper[Has0[String] & R]

where:    R is a type variable

      test(zz)

```

### LINKED ISSUE #18226: Type inference issues when using intersection types
## Compiler version

3.3.0

## Minimized code

```scala
trait F[-R]

trait Row[A]

def eliminateInt[R](f: F[R & Row[Int]]): F[R] = new F[R] {}

val x = new F[Row[Int] & Row[String]] {}

val _ = eliminateInt[Row[String]](x) // compiles OK when given explicit type
val _ = eliminateInt(x) // error!
```

## Output

```
Found:    (app.x : app.F[app.Row[Int] & app.Row[String]])
Required: app.F[app.Row[Int]]
```

## Expectation
I expect compiler to be able to infer `[R]` instantiation based on a type of a passed value.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #18226
Fixes #12077 

Might fix some other reported issues with AndTypes as well.

### CODE REVIEW

**Comment by guizmaii** [comment]
@odersky I see some `zio.Has` in the logs. Is the community build still using ZIO1? Because ZIO1 is not maintained anymore. That's not really interesting for you to test Dotty on it. If it's still ZIO1, I'd advise you guys to migrate to ZIO2 🙂

**Comment by smarter** [comment]
Minimized:
```scala
class Has[A]
trait Foo

class TestAspect[+LowerR, -UpperR]

class Spec[-R] {
  def foo[R1 <: R](aspect: TestAspect[R1, R1]): Unit = {}
}

class SuiteBuilder[R <: Has[_]] {
  def toSpec(
    spec: Spec[R & Has[Foo]],
    aspect: TestAspect[
      R & Has[Foo],
      R & Has[Foo]
    ]
  ) =
    spec.foo(aspect)
}
```

**Comment by NavidJalali** [comment]
> Thanks for the info! It would be great to get a PR submitting zio2 to the community build!

I'm gonna be honest, I just realized what the community build even is, but I might give it a shot, are these the steps? https://dotty.epfl.ch/docs/contributing/community-build.html

**Comment by bishabosha** [comment]
>  but I might give it a shot, are these the steps? https://dotty.epfl.ch/docs/contributing/community-build.html

yeah that's correct, you can also look at an old PR that added a project e.g. https://github.com/lampepfl/dotty/pull/14911/files

**Comment by odersky** [comment]
Thanks for the minimization @smarter. That illustrated another problem that was previously hidden by the wrong baseType handling.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -898,21 +898,43 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
       canWidenAbstract && acc(true, tp)
 
-    def tryBaseType(cls2: Symbol) = {
-      val base = nonExprBaseType(tp1, cls2).boxedIfTypeParam(tp1.typeSymbol)
+    def tryBaseType(cls2: Symbol) =
+
+      def computeBase(tp: Type): Type = tp.widenDealias match
+        case tp @ AndType(tp1, tp2) =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:904]: I would call the operands `tp11 and tp12`  to avoid shadowing the existing tp1/tp2.

**Comment by odersky** [comment]
That fixed it! In light of the problem it found we can still keep ZIO 1 in frozen state in the community build, until there's something new that would have to be changed in ZIO itself. But it would be good to also get ZIO 2.

**Review by dwijnand — APPROVED** [review]
Very nice.  I'd love a #how-i-fixed-it, if possible.

**Comment by odersky** [comment]
**How I  fixed it**

I compiled with -explain, which gave me a subtype trace, but that one came too late since the problematic type variable `R` in #18266 had already been instantiated to `Any`. So I turned all traces on in Typer and TypeComparer by changing occurrences of `trace` to `trace.force` in the source. That gave me a long log, and I searched for when the variable `x` was first typed and adapted. That pointed me to the interesting subtype trace that should have passed but did not. The goal that should have succeeded but did not was

     R & Row[Int] <: Row[String]

with type variable R uninstantiated. The goal should have succeeded in `fourthTry` by going through an `either` of two alternative tests. But it failed before getting there. I then added some debug printlns to TypeComparer to figure out which branches it took and after a while I realized that because the RHS is a named type `Row[String]` it tried to compute the base type of the LHS with class `Row`. That base type exists; it is `Row[Int]`. The logic then continued with that base type and failed since `Row[Int] !<: Row[String]`. I then tried to turn the base type logic off if the (widened, dealiased) LHS is an AndType. That made the test pass. 

But I knew that change would be too restrictive. We sometimes need the base type computation, as in here:
```scala
class Row[+X]
class A
class B
class C extends Row[A]
class D extends Row[B]
C & D <: Row[A & B]
```
Here we compute the base type of `Row` for `C & D`, which is Row[A & B], and succeed. But neither `C` nor `D` alone is a subtype of `Row[A & B]`.

So I then tried to identified exactly those cases where the base type would lose information, and proceed to `fourthTry` in those cases. I missed a case, which was exhibited in ZIO 1 and which @smarter minimized. That case was fixed by a slight change in [3a252a7](https://github.com/lampepfl/dotty/pull/18235/commits/3a252a7ec45393d99d3319252619892bb0db5d3c).

**Comment by odersky** [comment]
Writing my previous comment showed me that the logic can be simplified: We can always try to do the basetype computation, but if that fails we can see whether LHS was an AndType and proceed to fourthTry.

Thinking more about it ... On the other hand, if we do that, the computation with the approximated base type might instantiate type variables on the right, which then could be instantiated to something less precise than otherwise. So it's not a "try base type, if that fails "fourthTry", but a "try both base type and fourthTry" and pick the one that succeeds while instantiating the least amount of variables (or most amount for GADTs). Which is `either`.

**Review by smarter — APPROVED** [review]
LGTM, but it's not clear to me if this change can go in the LTS since the risk of regression is hard to estimate. In this week's dotty meeting we were leaning towards disallowing type inference changes in the LTS.

**Comment by odersky** [comment]
I also think backporting this to the LTS would be too risky.

**Comment by smarter** [comment]
The main branch is still 3.3.x, we haven't made a branch for the lts yet

**Pull request merged**